fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -212,23 +240,43 @@ def reduce(*_: Any) -> None:
 
                 # Make sure that this is not fired twice
                 self._grad_to_be_reduced[index] = False
-                param.grad.mul_(self.world_size_scaling)
-
-                # Future work includes clearing up the buffer if possible
-                def cleanup() -> None:
-                    if dst_rank != self.global_rank:
-                        param.grad = None
-
-                # Async reduce for this buffer, log the future
-                optimizer.work_handles.append(
-                    Workhandle(
-                        handle=dist.reduce(
-                            tensor=param.grad.data, dst=dst_rank, group=self.process_group, async_op=True
-                        ),
-                        callback=cleanup,
+
+                if not self.use_buckets or not self._should_bucket_grad[index]:
+                    param.grad.mul_(self.world_size_scaling)
+
+                    # Future work includes clearing up the buffer if possible
+                    def cleanup() -> None:
+                        if dst_rank != self.global_rank:
+                            param.grad = None
+
+                    # Async reduce for this buffer, log the future
+                    optimizer.work_handles.append(
+                        Workhandle(
+                            handle=dist.reduce(
+                                tensor=param.grad.data, dst=dst_rank, group=self.process_group, async_op=True
+                            ),
+                            callback=cleanup,
+                        )
                     )
-                )
-                self._reduced_grads[optimizer] += 1
+                    self._reduced_grads[optimizer] += 1
+                else:
+                    bucket = self.buckets[optimizer][param.device][dst_rank]
+                    bucket.params_checked_in += 1
+
+                    if bucket.full():
```

Comment:
will there be a chance the bucket is not full and reduce is called? if so, why? Do we have to deal with models that has unused params like find_unused_params in ddp?

Reply:
I think that replicating the DDP feature to find unused parameters would be a lot of work without any visible upside, it requires tracing the whole FW pass and it's already doable with DDP. The gist of the issue around that here is that there's no signal at the python level for when the BW pass is done, so no real way that I know of to catch that, and I've tried (cc @mrshenli, but this has been discussed extensively already).
If there's no unused parameter, then all the buckets will be flushed, I can add an assert in the next FW so that if it was not the case it can recommend to check for unused params ?

Reply:
understood. adding a comment would be good. can revisit later for sure

Reply:
adding an assert right now, it's a very good call I think, results would be wrong if this is missed